From 41f8162f18d2aa50cba8b1b19128751da3e90e10 Mon Sep 17 00:00:00 2001
From: Dennis Marttinen <twelho@welho.tech>
Date: Wed, 25 Mar 2026 16:27:18 +0200
Subject: [PATCH] md/bcache: prevent writeback from getting disabled as the
 cache fills

bcache will disable writeback once the cache fills beyond a certain threshold. The threshold cannot be beyond 70% (or 90% for sync writes), and the cache will get filled to (close to) 100% since nothing clears it until the freelists get too small (~0.1% of the cache). This means that writeback is effectively killed after not that much use, and stays so indefinitely. These changes aim to fix that by

- always forcing writeback
- setting the default writeback_percent to 50%
- disabling the bypass cutoff
- enlarging the freelists by 8x
---
 drivers/md/bcache/request.c   |  2 +-
 drivers/md/bcache/super.c     |  4 +++-
 drivers/md/bcache/writeback.c |  2 +-
 drivers/md/bcache/writeback.h | 25 ++++++-------------------
 4 files changed, 11 insertions(+), 22 deletions(-)

diff --git a/drivers/md/bcache/request.c b/drivers/md/bcache/request.c
index 3fa3b13a410f42..0aa9d623f00dbd 100644
--- a/drivers/md/bcache/request.c
+++ b/drivers/md/bcache/request.c
@@ -20,7 +20,7 @@
 
 #include <trace/events/bcache.h>
 
-#define CUTOFF_CACHE_ADD	95
+#define CUTOFF_CACHE_ADD	100
 #define CUTOFF_CACHE_READA	90
 
 struct kmem_cache *bch_search_cache;
diff --git a/drivers/md/bcache/super.c b/drivers/md/bcache/super.c
index 64bb38c9589565..589cc7b08309d8 100644
--- a/drivers/md/bcache/super.c
+++ b/drivers/md/bcache/super.c
@@ -2288,7 +2288,7 @@ static int cache_alloc(struct cache *ca)
 	 */
 	btree_buckets = clamp_t(size_t, ca->sb.nbuckets >> 7,
 				32, SB_JOURNAL_BUCKETS);
-	free = roundup_pow_of_two(ca->sb.nbuckets) >> 10;
+	free = roundup_pow_of_two(ca->sb.nbuckets) >> 7;
 	if (!free) {
 		ret = -EPERM;
 		err = "ca->sb.nbuckets is too small";
@@ -2944,6 +2944,8 @@ static int __init bcache_init(void)
 
 	bcache_is_reboot = false;
 
+	pr_info("[twelho] bcache patched: forced writeback, no bypass cutoff, enlarged freelists");
+
 	return 0;
 err:
 	bcache_exit();
diff --git a/drivers/md/bcache/writeback.c b/drivers/md/bcache/writeback.c
index 4b237074f453e3..8a2b38f2def9f9 100644
--- a/drivers/md/bcache/writeback.c
+++ b/drivers/md/bcache/writeback.c
@@ -1053,7 +1053,7 @@ void bch_cached_dev_writeback_init(struct cached_dev *dc)
 	dc->writeback_metadata		= true;
 	dc->writeback_running		= false;
 	dc->writeback_consider_fragment = true;
-	dc->writeback_percent		= 10;
+	dc->writeback_percent		= 50;
 	dc->writeback_delay		= 30;
 	atomic_long_set(&dc->writeback_rate.rate, 1024);
 	dc->writeback_rate_minimum	= 8;
diff --git a/drivers/md/bcache/writeback.h b/drivers/md/bcache/writeback.h
index 31df716951f66b..dcc6a81813adc5 100644
--- a/drivers/md/bcache/writeback.h
+++ b/drivers/md/bcache/writeback.h
@@ -2,11 +2,11 @@
 #ifndef _BCACHE_WRITEBACK_H
 #define _BCACHE_WRITEBACK_H
 
-#define CUTOFF_WRITEBACK	40
-#define CUTOFF_WRITEBACK_SYNC	70
+#define CUTOFF_WRITEBACK	100
+#define CUTOFF_WRITEBACK_SYNC	100
 
-#define CUTOFF_WRITEBACK_MAX		70
-#define CUTOFF_WRITEBACK_SYNC_MAX	90
+#define CUTOFF_WRITEBACK_MAX		100
+#define CUTOFF_WRITEBACK_SYNC_MAX	100
 
 #define MAX_WRITEBACKS_IN_PASS  5
 #define MAX_WRITESIZE_IN_PASS   5000	/* *512b */
@@ -102,27 +102,14 @@ extern unsigned int bch_cutoff_writeback_sync;
 static inline bool should_writeback(struct cached_dev *dc, struct bio *bio,
 				    unsigned int cache_mode, bool would_skip)
 {
-	unsigned int in_use = dc->disk.c->gc_stats.in_use;
-
 	if (cache_mode != CACHE_MODE_WRITEBACK ||
-	    test_bit(BCACHE_DEV_DETACHING, &dc->disk.flags) ||
-	    in_use > bch_cutoff_writeback_sync)
+	    test_bit(BCACHE_DEV_DETACHING, &dc->disk.flags))
 		return false;
 
 	if (bio_op(bio) == REQ_OP_DISCARD)
 		return false;
 
-	if (dc->partial_stripes_expensive &&
-	    bcache_dev_stripe_dirty(dc, bio->bi_iter.bi_sector,
-				    bio_sectors(bio)))
-		return true;
-
-	if (would_skip)
-		return false;
-
-	return (op_is_sync(bio->bi_opf) ||
-		bio->bi_opf & (REQ_META|REQ_PRIO) ||
-		in_use <= bch_cutoff_writeback);
+	return true;
 }
 
 static inline void bch_writeback_queue(struct cached_dev *dc)
